src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -479,34 +505,39 @@ private static boolean isDigit(int c) {
      * @return a decoded String
      */
     private String percentDecode(final String input) {
-        if (input == null) {
-            return null;
-        }
-        final String decoded = uriDecode(input);
-        if (!decoded.equals(input)) {
-            return decoded;
-        }
-        return input;
+        return uriDecode(input);
     }
 
     public static String uriDecode(String source) {
-        if (source == null) {
+        if (source == null || source.isEmpty()) {
             return source;
         }
-        int length = source.length();
-        StringBuilder builder = new StringBuilder();
+
+        if (source.indexOf('%') == -1) {
+            return source;
+        }
+
+        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
+        int length = bytes.length;
+        ByteArrayOutputStream buffer = new ByteArrayOutputStream(length);
+
         for (int i = 0; i < length; i++) {
```

Comment:
Minor optimization:
```suggestion
        int firstPercent = source.indexOf('%');
        if (firstPercent == -1) {
            return source;
        }

        byte[] bytes = source.getBytes(StandardCharsets.UTF_8);
        int length = bytes.length;
        ByteArrayOutputStream buffer = new ByteArrayOutputStream(length);

        for (int i = firstPercent; i < length; i++) {
```

Reply:
That won't work because unlike `StringBuilder`, `ByteArrayOutputStream` can't take the original byte array as an argument, unless I am missing something.